deno task dev fails on windows due to wrong paths

It fails when running the command 'deno task dev' after init the latest project.

--------------Error Message -------------------------
Task dev deno run -A --watch=static/,routes/ dev.ts
Watcher Process started.

Rebuilding...

Done in 211ms.
✘ [ERROR] Could not resolve "\Users\myname\AppData\Local\Temp\78ad55ed8210224/entrypoint.js" [plugin deno-resolver]

  The plugin "deno-resolver" didn't set a resolve directory, so esbuild did not search for
  "\Users\myname\AppData\Local\Temp\78ad55ed8210224/entrypoint.js" on the file system.

error: Uncaught (in promise) Error: Build failed with 1 error:
error: Could not resolve "\Users\myname\AppData\Local\Temp\78ad55ed8210224/entrypoint.js"
    at failureErrorWithLog (file:///C:/Users/myname/AppData/Local/deno/npm/registry.npmjs.org/esbuild/0.23.1/lib/main.js:1476:15)
    at file:///C:/Users/myname/AppData/Local/deno/npm/registry.npmjs.org/esbuild/0.23.1/lib/main.js:945:25
    at runOnEndCallbacks (file:///C:/Users/myname/AppData/Local/deno/npm/registry.npmjs.org/esbuild/0.23.1/lib/main.js:1316:45)
    at buildResponseToResult (file:///C:/Users/myname/AppData/Local/deno/npm/registry.npmjs.org/esbuild/0.23.1/lib/main.js:943:7)
    at file:///C:/Users/myname/AppData/Local/deno/npm/registry.npmjs.org/esbuild/0.23.1/lib/main.js:970:16
    at responseCallbacks.<computed> (file:///C:/Users/myname/AppData/Local/deno/npm/registry.npmjs.org/esbuild/0.23.1/lib/main.js:622:9)
    at handleIncomingPacket (file:///C:/Users/myname/AppData/Local/deno/npm/registry.npmjs.org/esbuild/0.23.1/lib/main.js:677:12)
    at Readable.readFromStdout (file:///C:/Users/myname/AppData/Local/deno/npm/registry.npmjs.org/esbuild/0.23.1/lib/main.js:600:7)
    at Readable.emit (ext:deno_node/_events.mjs:393:28)
    at addChunk (ext:deno_node/_stream.mjs:2438:16)
Watcher Process failed. Restarting on file change...

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -23,7 +23,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 - Make `window` available on server for registering web components
 - Set the response status property returned by a custom handler
-- Fix file paths on windows #29
+- Wrong file paths on windows #29
 
 ## [0.0.16] - 2024-11-12
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -23,6 +23,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 - Make `window` available on server for registering web components
 - Set the response status property returned by a custom handler
+- Fix file paths on windows #29
 
 ## [0.0.16] - 2024-11-12
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -13,7 +13,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ### Fixed
 
-## [0.0.17] - 2024-11-13
+## [0.0.17] - 2024-11-20
 
 ### Changed
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -13,6 +13,13 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ### Fixed
 
+## [0.0.17] - 2024-11-13
+
+### Fixed
+
+- Make `window` available on server for registering web components
+- Set the response status property returned by a custom handler
+
 ## [0.0.16] - 2024-11-12
 
 ### Fixed

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -15,6 +15,10 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [0.0.17] - 2024-11-13
 
+### Changed
+
+- BREAKING: Rename `AppTemplate` to `AppRoot`
+
 ### Fixed
 
 - Make `window` available on server for registering web components

--- a/deno.json
+++ b/deno.json
@@ -1,6 +1,6 @@
 {
   "name": "@limette/core",
-  "version": "0.0.16",
+  "version": "0.0.17",
   "exports": {
     ".": "./src/mod.ts",
     "./runtime/is-land.ts": "./src/runtime/is-land.ts",

--- a/init/deno.json
+++ b/init/deno.json
@@ -1,6 +1,6 @@
 {
   "name": "@limette/init",
-  "version": "0.0.16",
+  "version": "0.0.17",
   "exports": {
     ".": "./src/mod.ts"
   },

--- a/init/src/mod.ts
+++ b/init/src/mod.ts
@@ -1,7 +1,7 @@
 import { ensureDir } from "@std/fs/ensure-dir";
 
 // This value is changed in the release pipeline
-const LIMETTE_VERSION = "0.0.16";
+const LIMETTE_VERSION = "0.0.17";
 
 const projectName = prompt("Your project name?");
 

--- a/init/src/mod.ts
+++ b/init/src/mod.ts
@@ -122,10 +122,10 @@ customElements.define("island-counter", Counter);
 
 const _appRouteTs = `
 import { LitElement, html } from "lit";
-import type { AppTemplateOptions } from "@limette/core";
+import type { AppRootOptions } from "@limette/core";
 
 export default class App extends LitElement {
-  render(app: AppTemplateOptions) {
+  render(app: AppRootOptions) {
     return html\`<html>
       <head>
         <meta charset="utf-8" />

--- a/src/deps.ts
+++ b/src/deps.ts
@@ -15,7 +15,7 @@ export { parseImports } from "npm:parse-imports@2.2.1";
 export { DOMParser } from "jsr:@b-fuze/deno-dom@0.1.48";
 
 export { walk } from "jsr:@std/fs@1.0.5/walk";
-export { parse, join, isAbsolute, dirname } from "jsr:@std/path@1.0.8";
+export { parse, join, isAbsolute, dirname, toFileUrl, resolve, normalize, SEPARATOR } from "jsr:@std/path@1.0.8";
 export { emptyDir } from "jsr:@std/fs@1.0.5/empty-dir";
 export { ensureFile } from "jsr:@std/fs@1.0.5/ensure-file";
 export { encodeHex } from "jsr:@std/encoding@1.0.5/hex";

--- a/src/deps.ts
+++ b/src/deps.ts
@@ -24,5 +24,6 @@ export { encodeHex } from "jsr:@std/encoding@1.0.5/hex";
 export { unsafeHTML } from "lit/directives/unsafe-html.js";
 export { html, render } from "npm:@lit-labs/ssr@3.2.2";
 export { collectResult } from "npm:@lit-labs/ssr@3.2.2/lib/render-result.js";
+export { installWindowOnGlobal } from "npm:@lit-labs/ssr@3.2.2/lib/dom-shim.js";
 export { LitElementRenderer } from "npm:@lit-labs/ssr@3.2.2/lib/lit-element-renderer.js";
 export type { RenderInfo, RenderResult } from "npm:@lit-labs/ssr@3.2.2";

--- a/src/deps.ts
+++ b/src/deps.ts
@@ -1,24 +1,24 @@
-export { Application } from "jsr:@oak/oak@16.1.0/application";
-export { Router } from "jsr:@oak/oak@16.1.0/router";
-export { send } from "jsr:@oak/oak@16.1.0";
-export type { RouterContext } from "jsr:@oak/oak@16.1.0/router";
+export { Application } from "jsr:@oak/oak@17.1.3/application";
+export { Router } from "jsr:@oak/oak@17.1.3/router";
+export { send } from "jsr:@oak/oak@17.1.3";
+export type { RouterContext } from "jsr:@oak/oak@17.1.3/router";
 export type {
   Context,
   Response as OakResponse,
   Request as OakRequest,
-} from "jsr:@oak/oak@16.1.0";
+} from "jsr:@oak/oak@17.1.3";
 
-export { default as esbuild } from "npm:esbuild@0.23.1";
+export { default as esbuild } from "npm:esbuild@0.24.0";
 export { denoPlugins } from "jsr:@luca/esbuild-deno-loader@0.11.0";
-export { parseImports } from "npm:parse-imports@2.1.1";
+export { parseImports } from "npm:parse-imports@2.2.1";
 
 export { DOMParser } from "jsr:@b-fuze/deno-dom@0.1.48";
 
-export { walk } from "jsr:@std/fs@1.0.2/walk";
-export { parse, join, isAbsolute, dirname } from "jsr:@std/path@1.0.3";
-export { emptyDir } from "jsr:@std/fs@1.0.2/empty-dir";
-export { ensureFile } from "jsr:@std/fs@1.0.2/ensure-file";
-export { encodeHex } from "jsr:@std/encoding@1.0.3/hex";
+export { walk } from "jsr:@std/fs@1.0.5/walk";
+export { parse, join, isAbsolute, dirname } from "jsr:@std/path@1.0.8";
+export { emptyDir } from "jsr:@std/fs@1.0.5/empty-dir";
+export { ensureFile } from "jsr:@std/fs@1.0.5/ensure-file";
+export { encodeHex } from "jsr:@std/encoding@1.0.5/hex";
 
 // @ts-ignore lit is a npm package and Deno doesn't resolve the exported members
 export { unsafeHTML } from "lit/directives/unsafe-html.js";

--- a/src/deps.ts
+++ b/src/deps.ts
@@ -12,7 +12,7 @@ export { default as esbuild } from "npm:esbuild@0.23.1";
 export { denoPlugins } from "jsr:@luca/esbuild-deno-loader@0.11.0";
 export { parseImports } from "npm:parse-imports@2.1.1";
 
-export { DOMParser, Element } from "jsr:@b-fuze/deno-dom@0.1.48";
+export { DOMParser } from "jsr:@b-fuze/deno-dom@0.1.48";
 
 export { walk } from "jsr:@std/fs@1.0.2/walk";
 export { parse, join, isAbsolute, dirname } from "jsr:@std/path@1.0.3";

--- a/src/dev/build.ts
+++ b/src/dev/build.ts
@@ -12,6 +12,10 @@ import {
   emptyDir,
   ensureFile,
   join,
+  SEPARATOR,
+  resolve,
+  normalize,
+  toFileUrl
 } from "../deps.ts";
 import { fileExists } from "../server/utils.ts";
 import type { GetRouterOptions, Handlers } from "../server/router.ts";
@@ -78,7 +82,7 @@ async function buildJS(path: string, { devMode }: { devMode?: boolean }) {
   ].join("\n");
 
   const tempDirPath = await Deno.makeTempDir();
-  const entrypointPath = `${tempDirPath}/entrypoint.js`;
+  const entrypointPath = join(tempDirPath, '/entrypoint.js');
   await Deno.writeTextFile(entrypointPath, entryPoint, {});
 
   const esbuildResult = await esbuild.build({
@@ -88,7 +92,7 @@ async function buildJS(path: string, { devMode }: { devMode?: boolean }) {
         configPath: join(Deno.cwd(), "deno.json"),
       }),
     ],
-    entryPoints: [entrypointPath],
+    entryPoints: [toFileUrl(entrypointPath).href],
     sourcemap: devMode,
     minify: !devMode,
     bundle: true,
@@ -275,7 +279,7 @@ export async function getRoutes({
     const route: BuildRoute = {
       id,
       path,
-      relativeFilePath: `./${entry.path}`,
+      relativeFilePath:`.${SEPARATOR}${entry.path}`,
       absoluteFilePath,
       routeModule: (await loadFs?.(entry.path)) as { default: unknown },
       tagName,
@@ -373,7 +377,7 @@ export async function build() {
     }
 
     // Generate static routes file
-    routesImportsString += `import * as route${routeIndex} from ".${route.relativeFilePath}";
+    routesImportsString += `import * as route${routeIndex} from "${toFileUrl(route.absoluteFilePath)}";
 `;
 
     routesArrayString += `

--- a/src/dev/extract-islands.ts
+++ b/src/dev/extract-islands.ts
@@ -41,6 +41,8 @@ async function injectPatches(path: string) {
     // Import the DOM shim for SSR to provide customElements, window, and document
     import { installWindowOnGlobal } from "npm:@lit-labs/ssr/lib/dom-shim.js";
     installWindowOnGlobal();
+    // Set window object, because the shim doesn't do it
+    globalThis.window = globalThis;
   `;
 
   // **Step 2: Inject Patches for console and customElements.define**

--- a/src/dev/extract-islands.ts
+++ b/src/dev/extract-islands.ts
@@ -39,7 +39,7 @@ async function injectPatches(path: string) {
   // **Step 2: Inject DOM Shim Import from lit-labs/ssr**
   const domShimImport = `
     // Import the DOM shim for SSR to provide customElements, window, and document
-    import { installWindowOnGlobal } from "npm:@lit-labs/ssr/lib/dom-shim.js";
+    import { installWindowOnGlobal } from "npm:@lit-labs/ssr@3.2.2/lib/dom-shim.js";
     installWindowOnGlobal();
     // Set window object, because the shim doesn't do it
     globalThis.window = globalThis;

--- a/src/dev/path.ts
+++ b/src/dev/path.ts
@@ -1,4 +1,4 @@
-import { join, isAbsolute, dirname } from "../deps.ts";
+import { join, isAbsolute, dirname, toFileUrl } from "../deps.ts";
 
 let tailwindcss: string | undefined = undefined;
 
@@ -35,7 +35,7 @@ export async function resolvePath(
   // Step 4: If it's still a relative path, resolve it to an absolute path
   if (!isAbsolute(inputPath)) {
     const baseDir = dirname(basePath);
-    inputPath = join(baseDir, inputPath);
+    inputPath = toFileUrl(join(baseDir, inputPath)).href;
   }
 
   // Step 5: Return the resolved absolute path

--- a/src/mod.ts
+++ b/src/mod.ts
@@ -1,4 +1,4 @@
 export { LimetteApp } from "./server/app.ts";
 export { dev } from "./dev/dev.ts";
-export type { AppTemplateOptions } from "./server/ssr.ts";
+export type { AppRootOptions } from "./server/ssr.ts";
 export type { Handlers } from "./server/router.ts";

--- a/src/server/rendering/limette-element-renderer.ts
+++ b/src/server/rendering/limette-element-renderer.ts
@@ -2,7 +2,7 @@
 import { unsafeCSS } from "lit";
 // @ts-ignore lit is a npm package and Deno doesn't resolve the exported members
 import type { LitElement } from "lit";
-import { LitElementRenderer, render } from "../../deps.ts";
+import { LitElementRenderer } from "../../deps.ts";
 import type { RenderInfo, RenderResult } from "../../deps.ts";
 import type { BuildRoute } from "../../dev/build.ts";
 import type { ComponentContext } from "../router.ts";

--- a/src/server/router.ts
+++ b/src/server/router.ts
@@ -47,14 +47,14 @@ export type ComponentContext = {
 export async function getRouter(options: GetRouterOptions) {
   const router = new Router();
 
-  const [routes, AppTemplate] = await Promise.all([
+  const [routes, AppRoot] = await Promise.all([
     getRoutes(options),
     getAppTemplate(options),
   ]);
 
-  if (!AppTemplate) {
+  if (!AppRoot) {
     throw new Error(
-      "You need to create an AppTemplate (_app.ts/js) to render a page."
+      "You need to create an AppRoot (_app.ts/js) to render a page."
     );
   }
 
@@ -95,7 +95,7 @@ export async function getRouter(options: GetRouterOptions) {
                 }
 
                 const content = await renderContent(
-                  AppTemplate,
+                  AppRoot,
                   route,
                   routerContext,
                   data
@@ -124,7 +124,7 @@ export async function getRouter(options: GetRouterOptions) {
     // Default behaviour if no GET handler is provided
     if (route.routeModule?.default && !route.routeModule?.handler?.GET)
       router.get(route.path, async (routerContext) => {
-        const content = await renderContent(AppTemplate, route, routerContext);
+        const content = await renderContent(AppRoot, route, routerContext);
 
         routerContext.response.type = "text/html";
         routerContext.response.body = content;

--- a/src/server/router.ts
+++ b/src/server/router.ts
@@ -110,6 +110,7 @@ export async function getRouter(options: GetRouterOptions) {
             };
             const response = await handlerFn(ctx);
 
+            routerContext.response.status = response.status;
             routerContext.response.type = response.headers.get(
               "Content-Type"
             ) as string;

--- a/src/server/ssr.ts
+++ b/src/server/ssr.ts
@@ -12,6 +12,12 @@ import type { ComponentContext } from "./router.ts";
 
 import "../runtime/is-land.ts"; // should use limette?
 
+import { installWindowOnGlobal } from "npm:@lit-labs/ssr/lib/dom-shim.js";
+installWindowOnGlobal();
+// Set window object, because the shim doesn't do it
+// @ts-ignore some components use the `window` reference for registration process
+globalThis.window = globalThis;
+
 type Params = {
   [key: string]: string;
 };

--- a/src/server/ssr.ts
+++ b/src/server/ssr.ts
@@ -22,7 +22,7 @@ type Params = {
   [key: string]: string;
 };
 
-export type AppTemplateOptions = {
+export type AppRootOptions = {
   css: DirectiveResult<UnsafeHTMLDirective> | string;
   js: string[] | TemplateResult[] | DirectiveResult<UnsafeHTMLDirective>[];
   component: DirectiveResult<UnsafeHTMLDirective>;
@@ -35,7 +35,7 @@ export declare class ComponentCtxMixinInterface {
 
 export declare class AppTemplateInterface {
   prototype: {
-    render(app: AppTemplateOptions): TemplateResult;
+    render(app: AppRootOptions): TemplateResult;
   };
 }
 
@@ -137,7 +137,7 @@ const ComponentCtxMixin = (base: typeof LitElement) => {
 };
 
 export function bootstrapContent(
-  AppTemplate: AppTemplateInterface,
+  AppRoot: AppTemplateInterface,
   route: BuildRoute,
   componentContext: ComponentContext
 ) {
@@ -155,7 +155,7 @@ export function bootstrapContent(
     componentContext
   )}</script>`;
 
-  const appTemplateOptions: AppTemplateOptions = {
+  const appTemplateOptions: AppRootOptions = {
     css: route.cssAssetPath
       ? unsafeHTML(`<link rel="stylesheet" href="${route.cssAssetPath}" />`)
       : ``,
@@ -168,11 +168,11 @@ export function bootstrapContent(
     component: unsafeHTML(component),
   };
 
-  return AppTemplate.prototype.render(appTemplateOptions);
+  return AppRoot.prototype.render(appTemplateOptions);
 }
 
 export async function renderContent(
-  AppTemplate: AppTemplateInterface,
+  AppRoot: AppTemplateInterface,
   route: BuildRoute,
   routerContext: RouterContext<typeof route.path>,
   data?: ComponentContext["data"]
@@ -181,7 +181,7 @@ export async function renderContent(
 
   const result = render(
     await bootstrapContent(
-      AppTemplate as AppTemplateInterface,
+      AppRoot as AppTemplateInterface,
       route,
       componentContext
     ),

--- a/src/server/ssr.ts
+++ b/src/server/ssr.ts
@@ -1,11 +1,4 @@
-import {
-  html,
-  unsafeHTML,
-  render,
-  collectResult,
-  DOMParser,
-  Element,
-} from "../deps.ts";
+import { html, unsafeHTML, render, collectResult, DOMParser } from "../deps.ts";
 import type { RouterContext } from "../deps.ts";
 // @ts-ignore lit is a npm package and Deno doesn't resolve the exported members
 import type { LitElement, TemplateResult } from "lit";
@@ -40,7 +33,7 @@ export declare class AppTemplateInterface {
   };
 }
 
-function registerComponent(
+function registerRouteComponent(
   ComponentClass: CustomElementConstructor,
   tagName: string
 ) {
@@ -125,79 +118,6 @@ function moveLmtHeadElements(htmlString: string) {
   return doc.documentElement.outerHTML;
 }
 
-function moveLmtHeadElementsFirstLevel(htmlString: string) {
-  const parser = new DOMParser();
-  const doc = parser.parseFromString(
-    htmlString,
-    "text/html"
-  ) as unknown as Document;
-
-  const uniqueElements = new Map(); // Track the last unique element by key
-  let lastTitleElement = null; // Track the last <title> element found
-
-  function processLmtHeadElements(root: Document | DocumentFragment) {
-    // Find all <lmt-head> elements within the given root
-    const lmtHeadElements = Array.from(root.querySelectorAll("lmt-head"));
-
-    lmtHeadElements.forEach((wrapper) => {
-      Array.from(wrapper.children).forEach((child) => {
-        if (child.tagName === "TITLE") {
-          // Keep only the last <title> element found
-          lastTitleElement = child;
-        } else {
-          const key = child.getAttribute("key");
-
-          if (key) {
-            // Only keep the latest element with each key
-            uniqueElements.set(key, child);
-          } else {
-            // Treat children without a key as unique
-            uniqueElements.set(Symbol(), child);
-          }
-        }
-      });
-
-      // Remove the <lmt-head> wrapper from the DOM
-      wrapper.remove();
-    });
-  }
-
-  // Process <lmt-head> elements in the main DOM
-  processLmtHeadElements(doc);
-
-  // Process <lmt-head> elements within each declarative shadow DOM template
-  const templates = doc.querySelectorAll(
-    "template[shadowroot]"
-  ) as unknown as HTMLTemplateElement[];
-  templates.forEach((template) => {
-    processLmtHeadElements(template.content); // Directly use template.content without re-parsing
-  });
-
-  // Remove redundant elements in <head> that match keys in uniqueElements
-  Array.from(doc.head.querySelectorAll("[key]")).forEach((headElement) => {
-    const key = headElement.getAttribute("key");
-    if (key && uniqueElements.has(key)) {
-      headElement.remove();
-    }
-  });
-
-  // Remove any existing <title> elements in <head> if we have a new one
-  if (lastTitleElement) {
-    Array.from(doc.head.querySelectorAll("title")).forEach((title) =>
-      title.remove()
-    );
-    // Append the last <title> element found to <head>
-    doc.head.appendChild(lastTitleElement);
-  }
-
-  // Append the deduplicated elements to <head>
-  uniqueElements.forEach((element) => {
-    doc.head.appendChild(element);
-  });
-
-  return doc.documentElement.outerHTML;
-}
-
 const ComponentCtxMixin = (base: typeof LitElement) => {
   class ComponentCtxClass extends base {
     __ctx: ComponentContext = { params: {}, data: undefined };
@@ -220,7 +140,7 @@ export function bootstrapContent(
   const ComponentClass = ComponentCtxMixin(
     componentModule?.default as typeof LitElement
   );
-  const component = registerComponent(
+  const component = registerRouteComponent(
     ComponentClass as unknown as CustomElementConstructor,
     route.tagName
   );

--- a/src/server/ssr.ts
+++ b/src/server/ssr.ts
@@ -12,7 +12,7 @@ import type { ComponentContext } from "./router.ts";
 
 import "../runtime/is-land.ts"; // should use limette?
 
-import { installWindowOnGlobal } from "npm:@lit-labs/ssr/lib/dom-shim.js";
+import { installWindowOnGlobal } from "../deps.ts";
 installWindowOnGlobal();
 // Set window object, because the shim doesn't do it
 // @ts-ignore some components use the `window` reference for registration process

